### PR TITLE
Fix offset handling of String#rindex with Regex

### LINKED ISSUE #5590: `String#rindex` with Regex treats offset wrong
These calls should all be equivalent, yet `rindex(search : Regex, offset : Int)` behaves differently:

```crystal
"a/b".rindex(/\//, 2) # => nil
"a/b".rindex(/\//) # => 1
"a/b".rindex('/', 2) # => 1
"a/b".rindex('/') # => 1
"a/b".rindex("/", 2) # => 1
"a/b".rindex("/") # => 1
```

The `offset` parameter in all `rindex` methods should be interpreted as upper bound, i.e. it starts searching for the first instance from that position towards the beginning of the string.
In the overload accepting a regular expression, it is treated as lower bound. If no `offset` is specified, it's default value is `0` (should be `size`).

### PR DESCRIPTION
**Pull request by straight-shoota**
This also addas a few specs to ensure all variants of #rindex treat offset similarly.

Closes #5590

### CODE REVIEW

**Comment by straight-shoota** [comment]
@asterite yes, the loop could just break if `offset < match_data.begin`.

File: src/string.cr [hunk 1]
```diff
@@ -2738,12 +2738,11 @@ class String
   end
 
   # ditto
-  def rindex(search : Regex, offset = 0)
-    offset += size if offset < 0
```
Comment (RX14) [inline src/string.cr]: Why is this removed? You must support negative indexing from the end of the string.

Comment (straight-shoota) [reply src/string.cr]: The substring (`self[0..offset]`) handles negative indexes. There is no need to bother until we change that for a more performant implementation.

Comment (RX14) [reply src/string.cr]: So why are the specs for negative indexes removed, and why do we have

```
return nil unless 0 <= offset <= size
```

which always returns nil for negative indexes?

Comment (straight-shoota) [reply src/string.cr]: Argh, true that won't work.

Comment (straight-shoota) [reply src/string.cr]: I think I'll just replace the implementation as suggested above. And add more specs...

**Comment by straight-shoota** [comment]
Can we merge this?
